Auth0/OAuth2Grant.swift

```diff
@@ -67,50 +81,99 @@ struct PKCE: OAuth2Grant {
-                    callback(Result.failure(error: error))
-                } else {
-                    callback(result)
-                }
+        let idToken = values["id_token"]
+        let responseType = self.responseType
+        let authentication = self.authentication
+        let leeway = self.leeway
+        let nonce = self.defaults["nonce"]
+        let maxAge = self.maxAge
+        let isHybridFlow = responseType.contains(.idToken)
+        let verifier = self.verifier
+        let redirectUrlString = self.redirectURL.absoluteString
+        let clientId = authentication.clientId
+        let validatorContext = IDTokenValidatorContext(authentication: authentication,
+                                                       leeway: leeway,
+                                                       nonce: nonce,
+                                                       maxAge: maxAge)
+        validate(idToken: idToken, for: responseType, with: validatorContext) { error in
+            if let error = error { return callback(.failure(error: error)) }
+            authentication
+                .tokenExchange(withCode: code, codeVerifier: verifier, redirectURI: redirectUrlString)
+                .start { result in
+                    switch result {
+                    case .failure(let error as AuthenticationError) where error.description == "Unauthorized":
+                        // Special case for PKCE when the correct method for token endpoint authentication is not set (it should be None)
+                        let webAuthError = WebAuthError.pkceNotAllowed("Unable to complete authentication with PKCE. PKCE support can be enabled by setting Application Type to 'Native' and Token Endpoint Authentication Method to 'None' for this app at 'https://manage.auth0.com/#/applications/\(clientId)/settings'.")
+                        return callback(.failure(error: webAuthError))
+                    case .failure(let error): return callback(.failure(error: error))
+                    case .success(let credentials):
+                        guard !isHybridFlow else {
+                            let newCredentials = Credentials(accessToken: credentials.accessToken,
+                                                             tokenType: credentials.tokenType,
+                                                             idToken: idToken,
+                                                             refreshToken: credentials.refreshToken,
+                                                             expiresIn: credentials.expiresIn,
+                                                             scope: credentials.scope)
+                            return callback(.success(result: newCredentials))
+                        }
+                        // Code flow
```

Comment:
**Nit:** Technically** all of this is "code flow" of some kind.

Reply:
How would you rephrase the comment so it's more accurate and useful?

Reply:
Please feel free to add suggestions.